### ISSUE TITLE
StatefulSet volumeclaimtemplates added when base did not include it

### ISSUE THREAD

**Issue by msau42** [kubernetes-sigs/kustomize#504]
This might be caused to https://github.com/kubernetes-sigs/kustomize/pull/483

I have a StatefulSet spec that doesn't use volumeClaimTemplates, but when I kustomize it with common labels, it adds it to the spec, which causes it to fail validation:

My base spec:
```
kind: StatefulSet
apiVersion: apps/v1
metadata:
  name: test
spec:
  serviceName: "test"
  replicas: 1
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
        - name: nginx
          image: nginx
```

After kustomize:
```
kind: StatefulSet
apiVersion: apps/v1
metadata:
  name: test
spec:
  serviceName: "test"
  replicas: 1
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
        - name: nginx
          image: nginx
  volumeClaimTemplates:
    metadata:
      labels:
        app: my-label
```

Error:
```
error validating data: ValidationError(StatefulSet.spec.volumeClaimTemplates): invalid type for io.k8s.api.apps.v1.StatefulSetSpec.volumeClaimTemplates: got "map", expected "array"; if you choose to ignore these errors, turn validation off with --validate=false
```

**Comment by Liujingfang1** [comment]
I'll look into this issue. Before we have a fix, you can unblock yourself with following steps to modify the transformer configurations.
```
kustomize config save -d ~/.kustomize/config
```

In `~/.kustomize/config/commonlabel.YAML`, remove ollowing block or change create from true to false
```
-path: spec/volumeClaimTemplates/metadata/labels
  create: true
  group: apps
  kind: StatefulSet
```
When you run kustomize build, using `-t` to pick up the updated transformer configs.
```
kustomize build <die> -t ~/.kusotmize/config
```

**Comment by robbyt** [comment]
@Liujingfang1 Thank you for the workaround. I also have this same issue, after upgrading to release 1.0.9.

**Comment by twz123** [comment]
I also hit this one. Interestingly, kustomize replaces absent volumeClaimTemplates - which are usually an array - with a map. Note the missing hyphen in the above example. I think kustomize's transformer configs miss a distinction between something that is an array vs something that is an object.

In my use case, overriding transformer configs is not helping much, since I have a kustomization that has both statefulsets with and without volumeClaimTemplates. I use a patch instead that explicitly sets `volumeClaimTemplates: []`.

**Comment by turbotankist** [comment]
put 
```
spec:
 volumeClaimTemplates: []
```
in your StatefulSet yml

**Comment by schweikert** [comment]
The workaround described in https://github.com/kubernetes-sigs/kustomize/issues/504#issuecomment-433502027 doesn't work anymore with kustomize 2.0.3 because the -t option is not supported anymore. Also, it seems that it is not possible to override the configuration of the commonLabels transformer (using configurations:...)

The other workaround of using "volumeClaimTemplates: []" doesn't work for me because kubectl apply complains about a forbidden change.

In other words: we need this fix :) Right now we are stuck with kustomize 1.0.8

**Comment by gysel** [comment]
We're also struggling with this issue.

The workaround using `volumeClaimTemplates: []` kind of works but results in `for: "STDIN": StatefulSet.apps "xxx" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden.` for every deployment.

**Comment by Djabx** [comment]
The issue is always here.
/reopen

**Comment by HenriTEL** [comment]
As pointed out by @gysel you can set `volumeClaimTemplates: []` to make kustomize not generate a wrong StatefulSet.
You can try a `kubectl patch` to work around the `Forbidden: updates to statefulset spec` when the StatefulSet is already deployed.

**Comment by arianvp** [comment]
This issue is still very relevant though right? Kustomize generates a manifest that isn't even valid. Can we reopen this?

**Comment by Shell32-Natsu** [comment]
I cannot reproduce this issue with latest kustomize. If you are using kustomize shipped with kubectl, please take a look at #1500 and #2506.

**Comment by arianvp** [comment]
Yep you're right. I cant reproduce either on latest kustomize. Feel free to close this issue then in favour of the other two you just linked.
